Iterative [multigpu] adversarial training on MNIST and CIFAR10.

fartashf opened the pull request.

goodfeli commented on examples/multigpu_advtrain/README.md:
```diff
@@ -0,0 +1,6 @@
+Try these commands:
```
Clean up wording here

fartashf replied:
5268c83

goodfeli commented on examples/multigpu_advtrain/cifar_input.py:
```diff
@@ -0,0 +1,154 @@
+# From RevNets code
```
Should we put a specific URL?
Is the whole file from the other repo?
What license does the other repo use? We probably need to copy their copyright statement

fartashf replied:
License and url added in 7c0b864

goodfeli commented on examples/multigpu_advtrain/cifar_input.py:
```diff
@@ -0,0 +1,154 @@
+        # log.info("Apply random flipping")
+        image = tf.image.random_flip_left_right(image)
+    # Brightness/saturation/constrast provides small gains .2%~.5% on cifar.
+    # image = tf.image.random_brightness(image, max_delta=63. / 255.)
```
Let's remove commented code

fartashf replied:
Flag added in 0c492ac

goodfeli commented on examples/multigpu_advtrain/evaluator.py:
```diff
@@ -0,0 +1,333 @@
+import os
```
Add a docstring at the top of the file saying what it's for

fartashf replied:
Added in 1e805c5.

goodfeli commented on examples/multigpu_advtrain/evaluator.py:
```diff
@@ -0,0 +1,333 @@
+from target_attack import ProjectedGradientDescentMethod
+
+
+def model_eval(sess, x, y, predictions=None, X_test=None, Y_test=None,
```
Is it necessary to copy-paste this `model_eval` function or could we re-use the one from the core library?

fartashf replied:
The `model_eval` in the core library cannot handle a batch not divisible by batch_size. So this modification is good to have in the core `model_eval` is well. I made a new commit 2ef5a2c and changed the core `model_eval` and removed the one in this file.

goodfeli replied:
OK, thanks. For future work, it would be better to make the change to `model_eval` in a separate PR, so we could review one small PR updating `model_eval`. It gets difficult to review large PRs containing multiple changes.

goodfeli commented on examples/multigpu_advtrain/evaluator.py:
```diff
@@ -0,0 +1,333 @@
+    if model is None and predictions is None:
+        raise ValueError("One of model argument "
+                         "or predictions argument must be supplied.")
+    if model is not None:
```
If we do copy-paste the function, we shouldn't copy-paste the deprecated parts

fartashf replied:
Address by 2ef5a2c.

goodfeli commented on examples/multigpu_advtrain/evaluator.py:
```diff
@@ -0,0 +1,333 @@
+    return accuracy
+
+
+def get_attack(model, x, attack_type, *args, **kwargs):
```
what is the point of this function?

fartashf replied:
Clarified in f352aec.

goodfeli commented on examples/multigpu_advtrain/evaluator.py:
```diff
@@ -0,0 +1,333 @@
+        return get_attack_2(model, x, attack_type, *args, **kwargs)
+
+
+def get_attack_2(model, x, attack_type, sess, batch_size=None, y=None,
```
It would probably be better to just provide a few config files / config scripts for each dataset, rather than to have this function with if statements that need to be maintained

fartashf replied:
Simplified in 3312e05. There is now one small dictionary for all the shared parameters.

goodfeli commented on examples/multigpu_advtrain/evaluator.py:
```diff
@@ -0,0 +1,333 @@
+    return adv_x
+
+
+class Evaluator(object):
```
Add a class-level docstring saying what the class is for / documenting the parameters of the init method

fartashf replied:
Fixed in 5f7bc3e and 44356a0.

goodfeli commented on examples/multigpu_advtrain/evaluator.py:
```diff
@@ -0,0 +1,333 @@
+        self.eval_params = {'batch_size': batch_size}
+
+        # Load adversarial examples generated with the CleverHans tutorial
+        # model
```
Do they have to be generated by the tutorial or could they be any adversarial examples compatible with some format specification?

fartashf replied:
Removed the black box attack in 7fcfc4c. Added a TODO to do a better implementation.

goodfeli commented on examples/multigpu_advtrain/evaluator.py:
```diff
@@ -0,0 +1,333 @@
+        self.sum_op = tf.summary.merge_all()
+
+    def log_value(self, tag, val, desc='', print_stats=False,
+                  print_only=False):
```
docstring

fartashf replied:
Fixed in 5f7bc3e and 44356a0.

goodfeli commented on examples/multigpu_advtrain/evaluator.py:
```diff
@@ -0,0 +1,333 @@
+        self.log_value('test_accuracy_trans_advs', acc_bbox,
+                       'Accuracy on transferred')
+
+    def evaluate(self, inc_epoch=True):
```
docstring

The other eval methods are specific about what they're evaluating (`eval_trans_adv`) so I'm not quite sure what to expect when I see a generic `evaluate` method.

fartashf replied:
Fixed in 5f7bc3e and 44356a0.

goodfeli commented on examples/multigpu_advtrain/make_model.py:
```diff
@@ -0,0 +1,35 @@
+from resnet_tf import ResNetTF
+
+
+def make_madry(nb_classes=10, input_shape=(None, 28, 28, 1), **kwargs):
```
we already have the Madry Lab Challenge model in another example. Should we just import that one, or is this one different somehow?

fartashf replied:
This one uses the multigpu layers. Clarified in 8e181e5 and 0861a8b.

goodfeli commented on examples/multigpu_advtrain/manager.py:
```diff
@@ -0,0 +1,118 @@
+import tensorflow as tf
```
add a docstring at the top of the file saying what it's for

goodfeli commented on examples/multigpu_advtrain/manager.py:
```diff
@@ -0,0 +1,118 @@
+    return result_images
+
+
+class Manager(object):
```
docstring explaining what this class is / what it does.

I'm a little bit confused about why it has init and eval methods but doesn't seem to have train methods.

fartashf replied:
This class is now turned into TrainManager and TrainerMultiGPU and TrainerSingleGPU with train and eval functions. The refactoring is done in commits between 4bd545ec41930128bfbc397fd34699e3c9e2c0e2 and f75fc0d7d4c361685578685f302a522bbd4fb2fa.

goodfeli commented on examples/multigpu_advtrain/model.py:
```diff
@@ -0,0 +1,339 @@
+                           trainable=trainable)
+
+
+class MLP(object):
```
If this is straight from the tutorials, should we just import it from the tutorials?

fartashf replied:
Fixed in 714afc0. Now inherits from MLP in tutorial_models.

goodfeli commented on cleverhans/utils_tf.py:
```diff
@@ -265,15 +267,17 @@ def model_eval(sess, x, y, predictions=None, X_test=None, Y_test=None,
+            cur_corr_preds = correct_preds.eval(feed_dict=feed_dict)
 
-            accuracy += (cur_batch_size * cur_acc)
+            accuracy += cur_corr_preds[:cur_batch_size].sum()
```
Thanks, this looks like a better implementation.
Do you know if the old version was giving inaccurate values for any of the benchmarks?

fartashf replied:
The benchmarks might have been slightly affected. On MNIST, the test set size is 10,000 which is not divisible by the typical mini-batch size 128.

goodfeli commented on examples/multigpu_advtrain/model.py:
```diff
@@ -0,0 +1,339 @@
+        return sync_ops
+
+
+class MLP_probs(Model):
```
Is this just a version of MLP that hides every output except "probs"? Is there a reason we need that class? Can't we just call `get_probs` on a regular MLP?

fartashf replied:
Fixed in 714afc0. Now only one MLPnGPU is used.

goodfeli commented on examples/multigpu_advtrain/model.py:
```diff
@@ -0,0 +1,339 @@
+        return self.mlp.create_sync_ops(host_device)
+
+
+class Layer(object):
```
how hard would it be to use the Layer class from tutorial_models?

fartashf replied:
Fixed in 714afc0. Now LayernGPU inherits from Layer in tutorial_models.

goodfeli commented on examples/multigpu_advtrain/model.py:
```diff
@@ -0,0 +1,339 @@
+                return self.fprop_noscope(x, **kwargs)
+
+
+class Linear(Layer):
```
similarly, how hard would it be to use the one from the tutorials?

fartashf replied:
Fixed in 714afc0. Changed to LinearnGPU as it has to use the LayernGPUs get_variable.

goodfeli commented on examples/multigpu_advtrain/model.py:
```diff
@@ -0,0 +1,339 @@
+        return tf.matmul(x, self.W) + self.b
+
+
+class Conv2D(Layer):
```
similarly, how hard would it be to use the one from the tutorials?

fartashf replied:
Fixed in 714afc0. Changed to Conv2DnGPU as it has to use the LayernGPUs get_variable.

goodfeli commented on examples/multigpu_advtrain/model.py:
```diff
@@ -0,0 +1,339 @@
+                            (1,), self.padding) + self.b
+
+
+class ReLU(Layer):
```
similarly, how hard would it be to use the one from the tutorials?

fartashf replied:
Removed in 714afc0.

goodfeli commented on examples/multigpu_advtrain/model.py:
```diff
@@ -0,0 +1,339 @@
+        return tf.nn.relu(x)
+
+
+class Softmax(Layer):
```
similarly, how hard would it be to use the one from the tutorials?

fartashf replied:
Removed in 714afc0.

goodfeli commented on examples/multigpu_advtrain/model.py:
```diff
@@ -0,0 +1,339 @@
+        return tf.nn.softmax(x)
+
+
+class Flatten(Layer):
```
similarly, how hard would it be to use the one from the tutorials?

fartashf replied:
Removed in 714afc0.

goodfeli commented on examples/multigpu_advtrain/model.py:
```diff
@@ -0,0 +1,339 @@
+        return tf.reshape(x, [-1, self.output_width])
+
+
+class MaxPool(Layer):
```
(or maybe this one isn't in the tutorials?)

fartashf replied:
MaxPool does not exist in tutorial_models.

goodfeli commented on examples/multigpu_advtrain/resnet_tf.py:
```diff
@@ -0,0 +1,354 @@
+from __future__ import absolute_import
```
Is this file also from an external repo?

goodfeli replied:
should link to the source, reproduce the copyright / license statement, etc.

fartashf replied:
Fixed in 847490c.

goodfeli commented on examples/multigpu_advtrain/resnet_tf.py:
```diff
@@ -0,0 +1,354 @@
+        if self.Conv2D is None:
+            self.reuse = True
+        else:
+            # else my own conv is used and no reuse is needed
```
Reword this comment, it won't be clear who "my" refers to in the future

fartashf replied:
Fixed in d1169ec.

goodfeli commented on examples/multigpu_advtrain/run.py:
```diff
@@ -0,0 +1,317 @@
+    _data_path['cifar10'] = os.environ['CIFAR10_PARDIR']
+
+
+def preprocess_batch(images_batch, preproc_func=None):
```
There's another function with the same name in manager.py. How are they similar / different?

fartashf replied:
Fixed in 9078122.

goodfeli commented on examples/multigpu_advtrain/run.py:
```diff
@@ -0,0 +1,317 @@
+
+def model_train(sess, model, x_pre, x, y, predictions, X_train, Y_train,
+                predictions_adv=None, init_all=True, evaluate=None,
+                verbose=True, args=None, hparams=None):
```
I imagine this one really does have to be copy-pasted since the multi-GPU training logic is probably quite a lot different from the existing `model_train`, but it would be good for the docstring to mention specifically how this one is different.

fartashf replied:
This is now part of TrainManager and TrainerMultiGPU and TrainerSingleGPU with most of the code being shared. The refactoring is done in commits between 4bd545ec41930128bfbc397fd34699e3c9e2c0e2 and f75fc0d7d4c361685578685f302a522bbd4fb2fa.

goodfeli commented on examples/multigpu_advtrain/run.py:
```diff
@@ -0,0 +1,317 @@
+                adv_loss = build_train_op(y, predictions_adv)
+                loss = (loss + adv_loss) / 2
+
+    loss = tf.Print(loss, [loss])
```
I'd suggest removing this, or if you want to keep it, add a ', "loss"' argument so that users can tell what is being printed to STDOUT. If users just see numbers without context they won't know what they are.

fartashf replied:
Fixed in 05716f1.

goodfeli commented on examples/multigpu_advtrain/run.py:
```diff
@@ -0,0 +1,317 @@
+
+    loss = tf.Print(loss, [loss])
+
+    if hparams.model_type == 'resnet_tf':
```
It's not clear to me from just this context what is going on here---maybe add a few lines of comments to explain?

fartashf replied:
This is addressed in bf49322c by factoring out these lines into a specific build_train_op that handles different models and different losses.

goodfeli commented on examples/multigpu_advtrain/run_multigpu.py:
```diff
@@ -0,0 +1,78 @@
+"""
+Copy-paste of mnist_l1rbf.py
```
This comment is out of date

fartashf replied:
Fixed in 81f2acb.

fartashf replied:
Fixed in 81f2acb.

goodfeli commented on examples/multigpu_advtrain/svhn_input.py:
```diff
@@ -0,0 +1,85 @@
+# started from CIFAR10 in RevNets
```
this comment should be cleaned up

fartashf replied:
Fixed in 3c5de09.

goodfeli commented on examples/multigpu_advtrain/svhn_input.py:
```diff
@@ -0,0 +1,85 @@
+# started from CIFAR10 in RevNets
```
we already have a cleverhans/utils/utils_mnist.py, maybe this should be cleverhans/utils/utils_svhn.py

goodfeli replied:
Since this is a derivative work from cifar_input.py, it should probably have a copy of Mengye's license in it too

fartashf replied:
The license fixed in 3c5de09. The file name fixed in 1fcb2a4.

goodfeli commented on examples/multigpu_advtrain/cifar_input.py:
```diff
@@ -0,0 +1,180 @@
+"""
```
we already have a cleverhans/utils/utils_mnist.py, maybe this should be cleverhans/utils/utils_cifar.py

fartashf replied:
Fixed in 1fcb2a4.

goodfeli commented on examples/multigpu_advtrain/svhn_input.py:
```diff
@@ -0,0 +1,85 @@
+import os
+
+import numpy as np
+import tensorflow as tf  # NOQA
```
What does NOQA mean?

fartashf replied:
Fixed in 7581186.

goodfeli commented on examples/multigpu_advtrain/target_attack.py:
```diff
@@ -0,0 +1,257 @@
+
+class ProjectedGradientDescentMethod(Attack):
+
+    """
```
The docstring and the class name are out of sync

fartashf replied:
Fixed in 460d3f3.

goodfeli commented on examples/multigpu_advtrain/target_attack.py:
```diff
@@ -0,0 +1,257 @@
+from model import clone_variable, MLP_probs
+
+
+class ProjectedGradientDescentMethod(Attack):
```
How is this different from the MadyEtAl attack merged yesterday?

Since most of the attacks are some form of projected gradient descent, the name should be more specific to the Madry Lab Challenge

fartashf replied:
Fixed in 460d3f3. The name changed to MadryEtAlMultGPU.

goodfeli commented on examples/multigpu_advtrain/target_attack.py:
```diff
@@ -0,0 +1,257 @@
+            eta = eta * self.eps / norm
+        return eta
+
+    def PGDpgd(self, x, **kwargs):
```
What does this method name mean?

fartashf replied:
Fixed in 460d3f3. Changed to `attack` to make similar to the code in attack.py.

goodfeli commented on examples/multigpu_advtrain/target_attack.py:
```diff
@@ -0,0 +1,257 @@
+                                'clip_max': np.float32}
+        self.structural_kwargs = ['ord', 'nb_iter']
+
+        if not isinstance(self.model, Model):
```
Can't you use CallableModelWrapper?
https://github.com/tensorflow/cleverhans/blob/master/cleverhans/attacks.py#L254

fartashf replied:
Fixed in 460d3f3. Made the new class inherit from MadryEtAl.

goodfeli commented on examples/multigpu_advtrain/target_attack.py:
```diff
@@ -0,0 +1,257 @@
+        if not isinstance(self.model, Model):
+            self.model = MLP_probs(self.model)
+
+    def clip_eta(self, eta):
```
This is now a standalone function in utils_tf.py

fartashf replied:
Fixed in 460d3f3.

goodfeli commented on examples/multigpu_advtrain/target_attack.py:
```diff
@@ -0,0 +1,257 @@
+
+        return adv_x
+
+    def PGDpgd_singleStep(self, x, eta, y):
```
for this method name, I still don't know what "PGDpgd" means

fartashf replied:
Fixed in 460d3f3. It is now inherited from MadryEtAl as `attack_single_step`.

goodfeli commented on examples/multigpu_advtrain/target_attack.py:
```diff
@@ -0,0 +1,257 @@
+        eta = self.clip_eta(eta)
+        return x, eta, y
+
+    def PGDpgd_multigpu(self, x, **kwargs):
```
for this method name, I still don't know what "PGDpgd" means

fartashf replied:
Fixed in 460d3f3.

goodfeli commented on examples/multigpu_advtrain/target_attack.py:
```diff
@@ -0,0 +1,257 @@
+
+        return inputs, outputs
+
+    def PGDcl(self, x, **kwargs):
```
I don't know what this method name means either

goodfeli commented on examples/multigpu_advtrain/target_attack.py:
```diff
@@ -0,0 +1,257 @@
+                     ord=np.inf, clip_min=None, clip_max=None,
+                     y_target=None, pgd=True, multigpu=False, ngpu=1,
+                     **kwargs):
+        """
```
it looks like the docstring here is copy-pasted from another class and the new parameters like `pgd` aren't documented

fartashf replied:
Fixed in 460d3f3.

goodfeli commented on examples/multigpu_advtrain/trainer.py:
```diff
@@ -0,0 +1,211 @@
+    return inputs, outputs, sync_ops
+
+
+def model_train(trainer, manager):
```
there's more than one `model_train` in this PR, can you comment on the differences between them?

fartashf replied:
This is now part of TrainManager and TrainerMultiGPU and TrainerSingleGPU with most of the code being shared. The refactoring is done in commits between 4bd545ec41930128bfbc397fd34699e3c9e2c0e2 and f75fc0d7d4c361685578685f302a522bbd4fb2fa.

goodfeli commented on examples/multigpu_advtrain/trainer.py:
```diff
@@ -0,0 +1,211 @@
+import six
```
docstring at the top to explain what this file does

goodfeli replied:
There is actually still no docstring at the top

fartashf replied:
Fixed in aca841f.

goodfeli commented on examples/multigpu_advtrain/resnet_tf.py:
```diff
@@ -0,0 +1,354 @@
+
+        return train_op
+
+    # TODO(xpan): Consider batch_norm in contrib/layers/python/layers/layers.py
```
it probably doesn't make sense to preserve TODOs when duplicating open source code

fartashf replied:
Fixed in 45cdaf7.

npapernot commented:
```
======================================================================
FAIL: Test if pep8 is respected.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/test-environment/lib/python3.5/site-packages/nose-1.3.7-py3.5.egg/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/build/tensorflow/cleverhans/cleverhans/devtools/tests/test_format.py", line 45, in test_format_pep8
    raise AssertionError("PEP8 Format not respected")
AssertionError: PEP8 Format not respected
-------------------- >> begin captured stdout << ---------------------
/home/travis/build/tensorflow/cleverhans/examples/multigpu_advtrain/trainer.py:21:1: W391 blank line at end of file
/home/travis/build/tensorflow/cleverhans/examples/multigpu_advtrain/trainer_sgpu.py:67:45: E128 continuation line under-indented for visual indent
/home/travis/build/tensorflow/cleverhans/examples/multigpu_advtrain/trainer_sgpu.py:71:45: E128 continuation line under-indented for visual indent
/home/travis/build/tensorflow/cleverhans/examples/multigpu_advtrain/manager.py:238:5: E301 expected 1 blank line, found 0
/home/travis/build/tensorflow/cleverhans/examples/multigpu_advtrain/manager.py:271:1: W391 blank line at end of file
--------------------- >> end captured stdout << ----------------------
```

fartashf commented:
Thanks. This PR is unfinished. I'm resolving the issues.

On Sun, Oct 1, 2017 at 3:09 PM, Nicolas Papernot <notifications@github.com>
wrote:

> ======================================================================
> FAIL: Test if pep8 is respected.
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/home/travis/miniconda/envs/test-environment/lib/python3.5/site-packages/nose-1.3.7-py3.5.egg/nose/case.py", line 197, in runTest
>     self.test(*self.arg)
>   File "/home/travis/build/tensorflow/cleverhans/cleverhans/devtools/tests/test_format.py", line 45, in test_format_pep8
>     raise AssertionError("PEP8 Format not respected")
> AssertionError: PEP8 Format not respected
> -------------------- >> begin captured stdout << ---------------------
> /home/travis/build/tensorflow/cleverhans/examples/multigpu_advtrain/trainer.py:21:1: W391 blank line at end of file
> /home/travis/build/tensorflow/cleverhans/examples/multigpu_advtrain/trainer_sgpu.py:67:45: E128 continuation line under-indented for visual indent
> /home/travis/build/tensorflow/cleverhans/examples/multigpu_advtrain/trainer_sgpu.py:71:45: E128 continuation line under-indented for visual indent
> /home/travis/build/tensorflow/cleverhans/examples/multigpu_advtrain/manager.py:238:5: E301 expected 1 blank line, found 0
> /home/travis/build/tensorflow/cleverhans/examples/multigpu_advtrain/manager.py:271:1: W391 blank line at end of file
> --------------------- >> end captured stdout << ----------------------
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/cleverhans/pull/278#issuecomment-333399223>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AG-A-ZioBStjZBcV2jycUz2G3AnE1yXPks5sn-N4gaJpZM4PklaA>
> .
>

goodfeli commented:
Let me know when I should re-review

goodfeli commented:
I see you pushed a few commits, should I re-review?

fartashf commented:
I think I have addressed all the comments. However, the numbers I get don't
match what I expect. I'm debugging it. It would greatly help if you could
review. It might help in figuring out what causes the mismatch of numbers.

On Fri, Oct 13, 2017 at 5:11 PM, Ian Goodfellow <notifications@github.com>
wrote:

> I see you pushed a few commits, should I re-review?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/cleverhans/pull/278#issuecomment-336568588>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AG-A-RKuBiyQnDxeeQz78Dqx6XguWx4Mks5sr9IQgaJpZM4PklaA>
> .
>

fartashf commented:
I just removed `run.py`. There is now a single script `run_multigpu.py` that does both single GPU run and multi-GPU and uses the new abstraction. I hope you did not spend time on `run.py`.

fartashf commented:
@goodfeli This PR is ready for re-review. Thanks.

catherio commented:
@goodfeli - wanted to bring this to your attention that @fartashf is ready for a re-review on this.

goodfeli commented on cleverhans/attacks.py:
```diff
@@ -1002,7 +1013,7 @@ def __init__(self, model, back='tf', sess=None):
                                 'clip_min': np.float32,
                                 'clip_max': np.float32}
-        self.structural_kwargs = ['ord', 'nb_iter']
+        self.structural_kwargs = ['ord', 'nb_iter', 'rand_init']
```
The new `rand_init` param should be documented somewhere, probably in the `generate` method where `ord` is documented

fartashf replied:
Documented in a93dc6c.

goodfeli commented on cleverhans/attacks.py:
```diff
@@ -1115,11 +1127,14 @@ def attack(self, x, y):
+            eta = self.attack_single_step(x, eta, y)
 
         adv_x = x + eta
         if self.clip_min is not None and self.clip_max is not None:
```
This file looks OK once `rand_init` has a docstring entry. In the future, it would be a good idea to send things like the new `rand_init` parameter as a separate PR. It's almost trivial to review on its own, but now that it's in this PR, it won't get merged until everything else in the PR is merged.

fartashf replied:
Got it. I'll do so next time. Thanks.

goodfeli commented on examples/multigpu_advtrain/trainer_multigpu.py:
```diff
@@ -0,0 +1,122 @@
+from runner import RunnerMultiGPU
+
+
+class TrainerMultiGPU(TrainManager):
```
class-level docstring please

fartashf replied:
Fixed in aca841f.

goodfeli commented on examples/multigpu_advtrain/trainer_multigpu.py:
```diff
@@ -0,0 +1,122 @@
+import logging
```
module-level dosctring please

fartashf replied:
Fixed in aca841f and 7946c48.

goodfeli commented on cleverhans/utils_cifar.py:
```diff
@@ -0,0 +1,175 @@
+"""
```
This and the SVHN reader might be better off in the directory for the example, since we aren't offering a standard library-wide way to read all different datasets.

fartashf replied:
Fixed in cc8226d.

fartashf commented on examples/multigpu_advtrain/trainer_singlegpu.py:
```diff
@@ -0,0 +1,62 @@
+import logging
```
Fixed in aca841f and 7946c48.

fartashf commented:
Thanks for the re-review. Requested changes are now addressed. New Docstrings are added and svhn/cifar utils moved.

fartashf commented:
Thanks for the merge.

goodfeli commented:
Welcome!

On Fri, Jan 12, 2018 at 2:54 PM, Fartash Faghri <notifications@github.com>
wrote:

> Thanks for the merge.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/cleverhans/pull/278#issuecomment-357376823>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAXrGlRlQdefwtozEmGpRT-Uz-b2hbpPks5tJ-KUgaJpZM4PklaA>
> .
>
